homeassistant/components/synology_dsm/__init__.py

```diff
@@ -89,6 +80,9 @@ async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
         hass.config_entries.async_forward_entry_setup(entry, "sensor")
     )
 
+    if not entry.update_listeners:
+        entry.add_update_listener(_async_update_listener)
```

Comment:
Yeah I saw, but that means I should store a UNDO_UPDATE_LISTENER key in the entry, which will make a breaking change again.
And the `ConfigEntry` has the feature built in so why not using it?

I tested to add the integration on dev, then go to this branch and start again, totally working...
Try to fix to add listerners before setup, otherwise not sure if it's linked.

Reply:
Try again with this new commit

Reply:
Actually written like this in the doc : https://developers.home-assistant.io/docs/config_entries_options_flow_handler/#options-support

Reply:
Still seeing it after setting options ~~twice~~
```
2020-05-05 23:24:09 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry localhost for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 216, in async_setup
    hass, self
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 59, in async_setup_entry
    return await hass.data[DOMAIN].async_setup_entry(entry)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 163, in async_setup_entry
    raise ValueError("Config entry has already been setup!")
ValueError: Config entry has already been setup!

```

I hit the same problem.  Fixing the docs is on my TODO list:
https://github.com/home-assistant/core/pull/33986#issuecomment-612254511

Shouldn't be a breaking change

```
hass.data[DOMAIN][entry.entry_id] = {
   SYNO_API: api,
   UNDO_UPDATE_LISTENER: undo_listener
}
```

and then change all the

`api = hass.data[DOMAIN][entry.entry_id]` to `api = hass.data[DOMAIN][entry.entry_id][SYNO_API]`

Reply:
I understood the first time about setting options twice :stuck_out_tongue_closed_eyes: 

I just don't understand how you do that ?

Reply:
Having multiple same integration ?

Reply:
Change the scan interval option to 10m
Change the scan interval option to 5m

Wait a few seconds

Error appears in the log.

Reply:
I am not having this issue while changing multiple times the scan interval in the option.

Still did proposed code.